src/pushed_authorization.rs

```diff
@@ -1,35 +1,30 @@
-use std::collections::HashMap;
+use std::future::Future;
 
 use crate::{
     authorization::AuthorizationDetail,
+    credential::RequestError,
+    http_utils::{content_type_has_essence, MIME_TYPE_FORM_URLENCODED, MIME_TYPE_JSON},
     profiles::AuthorizationDetaislProfile,
-    proof_of_possession::{
-        ConversionError, ParsingError, ProofOfPossessionBody, VerificationError,
+    types::ParUrl,
+};
+use oauth2::{
+    http::{
+        header::{ACCEPT, CONTENT_TYPE},
+        HeaderValue, Method, StatusCode,
     },
+    AuthUrl, ClientId, CsrfToken, HttpRequest, HttpResponse, PkceCodeChallenge, RedirectUrl,
 };
-use oauth2::{AuthUrl, CsrfToken, PkceCodeChallenge};
-use openidconnect::{core::CoreErrorResponseType, IssuerUrl, StandardErrorResponse};
+use openidconnect::{core::CoreErrorResponseType, IssuerUrl, Nonce, StandardErrorResponse};
 use serde::{Deserialize, Serialize};
-use serde_json::json;
 
 #[derive(Clone, Debug, Deserialize, Serialize)]
 pub struct ParRequestUri(pub String);
 
 impl ParRequestUri {
-    pub fn new_random(num_bytes: u32) -> Self {
-        use base64::{
-            alphabet,
-            engine::{self, general_purpose},
-            Engine as _,
-        };
-        use rand::{thread_rng, Rng};
-
-        const CUSTOM_ENGINE: engine::GeneralPurpose =
-            engine::GeneralPurpose::new(&alphabet::URL_SAFE, general_purpose::NO_PAD);
-        let random_bytes: Vec<u8> = (0..num_bytes).map(|_| thread_rng().gen::<u8>()).collect();
+    pub fn new_random() -> Self {
         Self(format!(
             "urn:ietf:params:oauth:request_uri:{:}",
-            CUSTOM_ENGINE.encode(random_bytes)
+            Nonce::new_random().secret().clone()
```

Comment:
I think the function should accept a user-provided nonce to let them decide on the length

Reply:
I left it as an Option<Nonce>, then the user can provide or generate one.